ChatSecure/Classes/Controllers/OTRProtocolManager.m

```diff
@@ -206,17 +207,37 @@ - (void)goAwayForAllAccounts {
     }
 }
 
-- (void)disconnectAllAccountsSocketOnly:(BOOL)socketOnly {
+- (void)disconnectAllAccountsSocketOnly:(BOOL)socketOnly timeout:(NSTimeInterval)timeout completionBlock:(nullable void (^)())completionBlock
+{
     @synchronized (self) {
-        [self.protocolManagers enumerateKeysAndObjectsUsingBlock:^(id key, id <OTRProtocol> protocol, BOOL *stop) {
-            [protocol disconnectSocketOnly:socketOnly];
-        }];
+        dispatch_group_t group = dispatch_group_create();
+        for (NSObject<OTRProtocol> *manager in self.protocolManagers.allValues) {
+            if (manager.connectionStatus != OTRProtocolConnectionStatusDisconnected) {
+                dispatch_group_enter(group);
+                __block NSString *token = nil;
+                token = [manager addObserverForKeyPath:@"connectionStatus"
+                                               options:0
+                                                 block:^(NSString *keyPath, NSObject <OTRProtocol> *mgr, NSDictionary *change) {
+                                                     if (mgr.connectionStatus == OTRProtocolConnectionStatusDisconnected) {
+                                                         dispatch_group_leave(group);
+                                                         [mgr removeObserverForToken:token];
```

Comment:
This could cause an observer leak. You might want to try using Facebook's [`KVOController`](https://github.com/facebook/KVOController) which is already a dependency, because it makes KVO a little easier.

Reply:
Sorry, don't see the leak. `mgr` here is a local parameter, not a captured variable.

Reply:
What I'm trying to say is that if the `connectionStatus` doesn't equal `OTRProtocolConnectionStatusDisconnected`, the observer won't be removed. If the dispatch group wait timer expires without the observers being removed.. things might get weird. You could keep track of the tokens and then remove any non-removed observers after the dispatch_group_wait.

Reply:
In that case the disconnected status would be set next time the app switches to foreground and the observers would still be removed, but just to be safe and avoid confusion I changed it so all of them are always cleaned up in a single place.